pmlc/dialect/stripe/transforms.cc

```diff
@@ -0,0 +1,205 @@
+  }
+  con.erase();
+}
+
+void LimitLower(ParallelForOp op, size_t arg, int64_t val) {
+  val = std::min(std::max(val, int64_t(0)), op.getRange(arg));
+  auto body = &op.inner().front();
+  OpBuilder builder(op);
+  llvm::SmallVector<int64_t, 8> ranges;
+  for (size_t i = 0; i < op.ranges().size(); i++) {
+    ranges.push_back(i == arg ? op.getRange(i) - val : op.getRange(i));
+  }
+  op.getOperation()->setAttr("ranges", builder.getI64ArrayAttr(ranges));
+  builder.setInsertionPointToStart(body);
+  auto new_idx = builder.create<AffinePolyOp>(op.getLoc(), AffinePolynomial(val));
+  body->getArgument(arg)->replaceAllUsesWith(new_idx);
+  new_idx.setAttr("coeffs", builder.getI64ArrayAttr({1}));
+  new_idx.getOperation()->setOperands({body->getArgument(arg)});
+}
+
+void LimitUpper(ParallelForOp op, size_t arg, int64_t val) {
+  val = std::min(std::max(val, int64_t(0)), op.getRange(arg));
+  OpBuilder builder(op);
+  llvm::SmallVector<int64_t, 8> ranges;
+  for (size_t i = 0; i < op.ranges().size(); i++) {
+    ranges.push_back(i == arg ? val : op.getRange(i));
+  }
+  op.getOperation()->setAttr("ranges", builder.getI64ArrayAttr(ranges));
+}
+
+// 'Lifts' a constraint.  The idea here is more move 'most' of the
+//  checks out of a parallel for, and to split the cases up to allow
+//  more constraint rewrites, eventually resulting in (hopefully) no
+//  constraints in the fast path.
+//
+// More exactly, translates:
+//
+// PF [inner] {
+//   Pre
+//   Con ([outer_poly] + [inner_poly] + const) >= 0 {
+//     GE Case
+//   } else {
+//     LT Case
+//   }
+// }
+//
+// into:
+//
+// Con ([outer_poly] + min(inner_poly) + const) >= 0 {
+//   PF [inner] {
+//     Pre
+//     GE Case
+//   }
+// } else {
+//   Cons ([outer_poly] + max(inner_poly) + const >= 0) {
+//     PF [inner] {
+//       Pre
+//       Con ([outer_poly] + [inner_poly] + const) >= 0 {
+//         GE Case
+//       } else {
+//         LT Case
+//       }
+//     }
+//   } else {
+//     PF [inner] {
+//       Pre
+//       LT Case
+//     }
+//   }
+// }
+//
+
+void LiftConstraint(ParallelForOp pf) {
+  // Extract the constraint op
+  auto pf_block = &pf.inner().front();
+  auto con = mlir::dyn_cast<ConstraintOp>(*std::prev(pf_block->end(), 2));
+  // Get the complete polynomial, and split it into innner + outer
+  auto opoly = AffinePolynomial(con.input());
+  AffinePolynomial ipoly;
+  for (mlir::BlockArgument* arg : pf_block->getArguments()) {
+    auto it = opoly.terms.find(arg);
+    if (it != opoly.terms.end()) {
+      ipoly.terms.emplace(*it);
+      opoly.terms.erase(it);
+    }
+  }
+  // Compute the range of the inner polynomial
+  auto irange = AffineRange(ipoly);
+  // Make a builder aimed right before the original parallel-for
+  OpBuilder builder(pf.getOperation());
+  Location loc = pf.getLoc();
+  // Build the rewritten version
+  auto oc = builder.create<ConstraintOp>(loc, builder.create<AffinePolyOp>(loc, opoly + AffinePolynomial(irange.min)));
```

Comment:
Could we use slightly more intuitive variable names? Not sure what `oc` stands for, but a few more letters or even whole words would help.